### PR TITLE
do not page-cache EphemeralFile contents

### PR DESCRIPTION
**Pull request by problame**
## Problem

tl;dr: the caching of `EphemeralFile` in `PageCache` complicates the Pageserver code base to a degree that isn't worth it.

Before this PR, `PageCache` had three types of pages:

* `ImmutableFilePage`: caches Delta / Image layer file contents
* `MaterializedPage`: caches results of Timeline::get (page materialization)
* `EphemeralPage`: caches `EphemeralFile` contents

`EphemeralPage` is quite different from `ImmutableFilePage` and `MaterializedPage`:

* Immutable and materialized pages are for the acceleration of (future) reads of the same data using `PAGE_CACHE_SIZE * PAGE_SIZE` bytes of DRAM.
* Ephemeral pages are a write-back cache of `EphemeralFile` contents.

`EphemeralFile` is only used by `InMemoryLayer`, for the following purposes:
  * write: when filling up the `InMemoryLayer`, via `impl BlobWriter for EphemeralFile`
  * read: when doing page reconstruction for a page@lsn that isn't written to disk 
  * read: when writing L0 layer files, we re-read the `InMemoryLayer` and put the contents into the L0 delta writer (`create_delta_layer`). This happens every 10min or when InMemoryLayer reaches 256MB in size.

Let's look at what advantages the page cache provided for the `InMemoryLayer` use case before this PR:

* about the **write path**:
  * the writes happen one `Value` at a time, append-only
  * the common case is that `Value` is a WAL record, and if it isn't a full-page-image WAL record, then it's smaller than `PAGE_SIZE`
  * So, the `EphemeralPage` pages act as a buffer for these `< PAGE_CACHE` sized writes.
  * If there's no page cache eviction between subsequent `InMemoryLayer::put_value` calls, the `EphemeralPage` is still resident, so the page cache avoids doing a `write` system call.
    * In practice, a busy page server will have page cache evictions because we only configure 64MB of page cache size.
  * TODO: quantify the above.
* **reads for page reconstruction**: they are rare because recently written pages are in compute page cache. So, the page cache provides little value.
* **reads for `create_delta_layer`**:
  * in classical LSM terms, this function is what writes the memory-resident `C0` tree into the disk-resident `C1` tree
  * in our system, though, the values of InMemoryLayer are stored in an EphemeralFile, and hence they are not guaranteed to be memory-resident. 
  * the access pattern is
    * read `Value`s in `Key, LSN` order, which is `!=` insert order
    * worst case: each `Value` is on a different page; hence the `BlockCursor`'s 1-page-sized buffer is ineffective
    * best case: subsequent values are located on the same page, hence `BlockCursor`s buffer is maximally effective
    * even in the best case, we access `256MB/PAGE_SIZE` pages in potentially random order.
  * assuming the worst case, the perf advantage of page cache is 0 or negative.
  * TODO: quantify the average case

## Summary of changes

Based on the observations outlined above, this PR makes the following changes:

* Rip out `EphemeralPage` from `page_cache.rs`
* Replace `ephemeral_file.rs`'s usage of `PageReadGuard` and `PageWriteGuard` with a new `DirtyBuffer` type.
* `DirtyBuffer` gets `PAGE_SIZE`d from a global `buffer_pool`. That `buffer_pool` is currently just the `malloc` but can be changed to a size-capped pool with a wait list in the future.
* `DirtyBuffer`'s `Drop` impl panics if the buffer hasn't been written back.

How does the above address **affect performance? It makes it worse**, but the advantage of not having page-cached EphemeralFile is worth it:

* **write path**: each `put_value` call becomes a `read` and one or more `write` system calls. The `read` to load the last page (it may be less-than-full), `write`s to append new values. Note that both read and write are likely to hit the kernel page cache and don't do fsync or similar. So, these system calls are likely just CPU overhead. We can reduce this overhead in the future by exploiting the fact that the write path is append-only. We actually don't need to `read` and `write` full pages; we can just use `write` to append to the file.
* **reads for page reconstruction**: we always do an 8K read system call. The system call will likely hit the kernel page cache because we wrote the data within the last 10min. So, the perf penalty is only the syscall overhead. We can avoid a full 8k read and use `preadv` in the future.
* **reads for `create_delta_layer`**: 
  * We do a `read()` system call for every value for now
    * Note that these reads are very likely to hit the kernel page cache, in which case they're not blocking
  * we have options to improve this in the future in a way that is more efficient than the current page_cache-based approach:
    * idea 1: prefetch pipeline + using `preadv` to avoid full page reads
    * idea 2: use `copy_file_range` to copy from ephemeral file into the L0 delta file

Lastly, with this PR, the page cache now holds just `ImmutableFilePage` and `MaterializedPage`.
As stated earlier, these types of pages are read-only.
Many great open-source libraries solve this problem better than our `page_cache.rs` (e.g., better replacement policy, less global locking).
We can now explore using them.

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

**Review by arpad-m — COMMENTED** [review]
The proposal itself looks great. It does require a new enum variant in the `BlockLease` enum introduced by #4982, which means the enum is now not zero-overhead outside of test cases. But given that this is code that is adjacent to I/O, the overhead shouldn't be that heavy.

The PR also makes my life easier with the async conversion, as I now only have to take care of immutable pages.

**Comment by hlinnaka** [comment]
How big is the performance impact?

> write path: each put_value call becomes a read and one or more write system calls. The read to load the last page (it may be less-than-full), writes to append new values. Note that both read and write are likely to hit the kernel page cache and don't do fsync or similar. So, these system calls are likely just CPU overhead. We can reduce this overhead in the future by exploiting the fact that the write path is append-only. We actually don't need to read and write full pages; we can just use write to append to the file.

That worries me from a performance point of view. Instead of writing a few bytes in memory, you need to do two system calls. System calls are cheap when the page is in kernel page cache, but it's still an order of magnitude slower. This needs to be benchmarked.

The write path is append-only, so we could have a special case for the last page that is still being modified, and to use the page cache for all the other ephemeral file pages which are read-only. I think you could just always keep the last page in memory. That'd be 8k of memory for each timeline, which is not insignificant, but seems acceptable.

> Replace ephemeral_file.rs's usage of PageReadGuard and PageWriteGuard with a new DirtyBuffer type.
> DirtyBuffer gets PAGE_SIZEd from a global buffer_pool. That buffer_pool is currently just the malloc but can be changed to a size-capped pool with a wait list in the future.
> DirtyBuffer's Drop impl panics if the buffer hasn't been written back.

I don't understand this new buffer pool. When you're modifying a page, you always read the page from the kernel, modify it, and write it back, right? What's the purpose of the pool then? You could just allocate a temporary 8k buffer on the spot, and free it when you're done. Is it just to reduce the malloc+free overhead?

**Comment by problame** [comment]
> The write path is append-only, so we could have a special case for the last page that is still being modified, and to use the page cache for all the other ephemeral file pages which are read-only. I think you could just always keep the last page in memory. That'd be 8k of memory for each timeline, which is not insignificant, but seems acceptable.

Yeah, that's nice.
Let me paraphrase: we will use `ImmutablePage` for all pages of EphemeralFile except the last page, which won't be page-cached but be a buffer inside EphemeralFile.

Correctly understood?

>  I don't understand this new buffer pool. When you're modifying a page, you always read the page from the kernel, modify it, and write it back, right? What's the purpose of the pool then? You could just allocate a temporary 8k buffer on the spot, and free it when you're done. Is it just to reduce the malloc+free overhead?

IIUC one of the points of page cache was to control memory usage. Hence I thought I'd prepare the way for a buffer pool that is size-limited.
But, I guess that branch of the conversation is irrelevant if we move to your proposal that I paraphrased above?

**Comment by hlinnaka** [comment]
> > The write path is append-only, so we could have a special case for the last page that is still being modified, and to use the page cache for all the other ephemeral file pages which are read-only. I think you could just always keep the last page in memory. That'd be 8k of memory for each timeline, which is not insignificant, but seems acceptable.
> 
> Yeah, that's nice. Let me paraphrase: we will use `ImmutablePage` for all pages of EphemeralFile except the last page, which won't be page-cached but be a buffer inside EphemeralFile.
> 
> Correctly understood?

Exactly!

> > I don't understand this new buffer pool. When you're modifying a page, you always read the page from the kernel, modify it, and write it back, right? What's the purpose of the pool then? You could just allocate a temporary 8k buffer on the spot, and free it when you're done. Is it just to reduce the malloc+free overhead?
> 
> IIUC one of the points of page cache was to control memory usage. Hence I thought I'd prepare the way for a buffer pool that is size-limited. But, I guess that branch of the conversation is irrelevant if we move to your proposal that I paraphrased above?

Yep.

**Comment by problame** [comment]
Superseded by https://github.com/neondatabase/neon/pull/4994
